File: Data Structures & Algorithms/three-integer-sum/submission-2.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> sol = {};

        for (int i=0; i < nums.size()-2; ++i) {
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            if (nums[i] > 0) break;
            
            int target = -nums[i];
            int l = i+1, r = nums.size()-1;
            //binary search
            while (l < r) {
                int curr_sum = nums[l] + nums[r];
                if (curr_sum < target) {
                    l += 1;
                } else if (curr_sum > target) {
                    r -= 1;
                } else {
                    sol.push_back({nums[i], nums[l], nums[r]});

                    //find more for same i
                    l += 1;
                    r -= 1;
                    while (l < r && nums[l] == nums[l-1]) l++;
                    while (l < r && nums[r] == nums[r+1]) r--;
                }
            }
        }
        return sol;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.cs
public class Solution {
    public int Search(int[] nums, int target) {
        int l = 0, r = nums.Length-1, m;
        while (l <= r) {
            m = l + (r-l)/2;
            if (target == nums[m]) return m;
            if (nums[l] <= nums[m]) {
                if (nums[l] <= target && target < nums[m]) {
                    r = m-1;
                } else {
                    l = m+1;
                }
            }
            else {
                if (nums[m] < target && target <= nums[r]) {
                    l = m+1;
                } else {
                    r = m-1;
                }
            }
        }
        return -1;
    }
}
